File: hw3/prob1/fraction.cc
/* NAME
 *  fraction.cc
 *
 * DESCRIPTION
 *  Defines methods for fraction class
 *
 * PARAMETERS
 *  See specific method
 *
 * RETURN
 *  See specific method
 * 
 * EXAMPLE
 *  See specific method
 *
 * NOTES
 *  Make sure the makefile is available to compile the code
 *
 * Alex Z. Chen - 10/18/2016
 * alexac9@uw.edu
 */
#include <iostream>
#include "fraction.hh"

fraction::fraction(int n, int d) : numerator(n), denominator(d) {
  if (0 == d) {
    throw fraction_exception("Attempted to assign 0 to the denominator.");
  } 
}
fraction::~fraction() {}
fraction::fraction(const fraction &f) {
  numerator = f.numerator;
  denominator = f.denominator;
}
fraction &fraction::operator=(const fraction &other) {
  if (this != &other) {
    numerator = other.numerator;
    denominator = other.denominator;
  }
  return *this;
}
void fraction::set_den(int d) {
  if (0 == d) {
    throw fraction_exception("Attempted to assign 0 to the denominator.");
  }
  denominator = d;
}
std::ostream& operator<<(std::ostream& os, const fraction &f) {
  os << f.num() << '/' << f.den();
  return os;
}

File: hw9/src/fraction.rs
//! NAME
//!  fraction.rs
//!
//! DESCRIPTION
//!  Module file that defines the fraction class
//!
//! PARAMETERS
//!  See specific method
//!
//! RETURN
//!  See specific method
//! 
//! EXAMPLE
//!  See specific method
//!
//! NOTES
//!  Make sure the Cargo.toml file is available to compile the code
//!
//! Alex Z. Chen - 12/05/2016
//! alexac9@uw.edu
//!

#[derive(Clone)]
// Data definition
pub struct Fraction {
  numerator: isize,
  denominator: isize,
}
// Constructors
pub fn fraction(n: isize, d: isize) -> Fraction {
  if 0 == d {
    panic!("Attempted to set denominator to 0");
  } 
  Fraction {numerator: n, denominator: d}
}
// Methods
impl Fraction {
  // Getters
  pub fn num(&self) -> isize { self.numerator }
  pub fn den(&self) -> isize { self.denominator }

  // Setters
  pub fn set_num(&mut self, n: isize) { self.numerator = n; }
  pub fn set_den(&mut self, d: isize) {
    if 0 == d {
      panic!("Attempted to set denominator to 0");
    }
    self.denominator = d;
  }

  // Reduce method
 pub fn reduce(&self) -> Fraction {
    let mut abs_num = self.num().abs();
    let mut abs_den = self.den().abs();
    let mut gcd = 1;

    // Modified Euclidean algorithm to find greatest common divisor (GCD)
    while (0 != abs_num) && (0 != abs_den) {
      if abs_num > abs_den {
        gcd = abs_den;
        abs_num %= abs_den;
      } else {
        gcd = abs_num;
        abs_den %= abs_num;
      }
    }
    // Eliminate double negatives or move negative to numerator
    if 0 > self.den() {
        gcd *= -1;
      }
    Fraction {numerator: self.num()/gcd, denominator: self.den()/gcd}
  }
}
